app/services/ClientService.scala

```diff
@@ -137,105 +122,83 @@ class ClientServiceImpl @Inject()(
 
   // getClients should be getAllClients or getUnassignedClients NOT getClients (maybe filtered)
   def saveSelectedOrFilteredClients(arn: Arn)
                                    (formData: AddClientsToGroup)
                                    (getClients: Arn => Future[Option[Seq[DisplayClient]]])
                                    (implicit hc: HeaderCarrier, ec: ExecutionContext, request: Request[Any]): Future[Unit] = {
 
-    formData.submit.trim match {
-      case CLEAR_BUTTON =>
-        for {
-          clients <- lookupClients(arn)(formData.clients)
-          _ <- addSelectablesToSession(clients.map(_.map(dc => dc.copy(selected = true)))
-          )(SELECTED_CLIENTS, FILTERED_CLIENTS)
-          _ <- clearSessionForSelectingClients()
-        } yield ()
-
-      case CONTINUE_BUTTON =>
-        for {
-          clients <- lookupClients(arn)(formData.clients)
-          _ <- addSelectablesToSession(
-            clients.map(_.map(_.copy(selected = true))))(
-            SELECTED_CLIENTS,
-            FILTERED_CLIENTS
-          )
-          _ <- clearSessionForSelectingClients()
-        } yield ()
-
-      case FILTER_BUTTON =>
-        if (formData.search.isEmpty && formData.filter.isEmpty) {
-          clearSessionForSelectingClients()
-        } else {
+    val selectedClientIds = formData.clients.getOrElse(Seq.empty)
+    val clientsMarkedSelected = for {
+      allClients <- getClients(arn)
+      selectedClients = allClients.fold(Option.empty[List[DisplayClient]])(clients =>
+        Some(clients.filter(cl => selectedClientIds.contains(cl.id)).map(_.copy(selected = true)).toList))
+      _ <- addSelectablesToSession(selectedClients)(SELECTED_CLIENTS, FILTERED_CLIENTS)
+    } yield allClients
+    clientsMarkedSelected.map { clients =>
+      formData.submit.trim match {
+        case FILTER_BUTTON =>
+          if (formData.search.isEmpty && formData.filter.isEmpty) {
+            clearSessionForSelectingClients()
+          } else {
+            for {
+              _ <- sessionCacheRepository.putSession(CLIENT_FILTER_INPUT, formData.filter.getOrElse(""))
+              _ <- sessionCacheRepository.putSession(CLIENT_SEARCH_INPUT, formData.search.getOrElse(""))
+              _ <- filterClients(formData)(clients)
+            } yield ()
+          }
+        case _ =>
           for {
-            clients <- lookupClients(arn)(formData.clients)
-            _ <- addSelectablesToSession(
-              clients.map(_.map(_.copy(selected = true)))
-            )(SELECTED_CLIENTS, FILTERED_CLIENTS)
-            _ <- sessionCacheRepository.putSession(CLIENT_FILTER_INPUT, formData.filter.getOrElse(""))
-            _ <- sessionCacheRepository.putSession(CLIENT_SEARCH_INPUT, formData.search.getOrElse(""))
-            _ <- filterClients(arn)(formData)(getClients)
+            _ <- sessionCacheRepository.deleteFromSession(FILTERED_CLIENTS)
```

Comment:
not sure if you could replace these lines with `clearSessionForSelectingClients()` or if it's better implemented as is